shotover/src/transforms/query_counter.rs

```diff
@@ -26,13 +30,23 @@ pub struct QueryCounterConfig {
 
 impl QueryCounter {
     pub fn new(counter_name: String) -> Self {
-        counter!("shotover_query_count", "name" => counter_name.clone());
+        // Leaking here is fine since the builder is created only once during shotover startup.
+        let counter_name_ref: &'static str = counter_name.leak();
+
+        // Although not incremented, this counter needs to be created to ensure shotover_query_count is 0 on shotover startup.
+        counter!("shotover_query_count", "name" => counter_name_ref);
 
         QueryCounter {
-            // Leaking here is fine since the builder is created only once during shotover startup.
-            counter_name: counter_name.leak(),
+            counter_name: counter_name_ref,
+            query_to_counter: HashMap::new(),
         }
     }
+
+    fn increment_counter(&mut self, query: Cow<str>, query_type: &'static str) {
+        self.query_to_counter.entry(query.to_string())
```

Comment:
It's interesting that `HashMap.entry()` doesn't support lookup by reference. I didn't go through the entire discussion in https://github.com/rust-lang/rfcs/pull/1769 but it kind of makes sense as we may need to insert key so that it needs to be owned?

Reply:
ah yeah, thats unfortunate.

Given that we need to convert the string to owned to use entry, the cow isnt actually giving us any advantage, we should revert the cow changes and go back to:
```rust
fn increment_counter(&mut self, query: String, query_type: &'static str) {
```

To do that, we also need to swap `or_insert_with` to `or_insert_with_key` so that we access the key while creating:
```rust
.or_insert_with_key(|query| counter!("shotover_query_count", "name" => self.counter_name, "query" => query.to_string(), "type" => query_type))
```

## Alternatively

We could alternatively keep the cow and revert the move to the `entry` API.
This should be faster since I expect allocating to be slower than hashing.
However its impossible to tell without implementing both approaches and benchmarking.

## But ultimately

But ultimately we are working on this to resolve a warning, so we shouldn't get too bogged down into these optimizations.
So I think just pick one of these solutions and go with it.
I'm happy with whichever one you choose.